### ISSUE TITLE
Slow partition spreading

### ISSUE THREAD

**Issue by pveentjer** [hazelcast/hazelcast#8213]
When a high number of partitions is selected, partitions could be balanced extremely unfair. This can lead to some members getting almost all partitions, and others nothing. I added some test case to a 4 node hz cluster with 12k partitions. This is the output of one of the members that didn't receive any partitions:

```
ERROR 13:52:59 Failure #2 C_A4_W1 AtomicLongTest WORKER_EXCEPTION[java.lang.IllegalStateException: No partitions found, partitionsPerMember:{Member [10.212.1.111]:5701 - 75fd59d0-137a-4fcd-8039-d6de95553ea0=11986, Member [10.212.1.113]:5701 - 9d3e70a8-caa5-411a-b41f-79834b45d6b7=14}]
```

As one can see member [10.212.1.111]:5701 has 11986 partitions, member [10.212.1.113]:5701 has 14 and the other remaining members have zero partitions. 

Perhaps this is caused by member Member [10.212.1.111]:5701 being available first and therefor it pulls all partitions to himself. Once the remaining member start, there is no partitons left for them. 

@mdogan will this imbalance be restored by the system?

**Comment by pveentjer** [comment]
A reproducer

```
package com.hazelcast;

import com.hazelcast.config.Config;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.Member;
import com.hazelcast.core.Partition;
import com.hazelcast.nio.Address;
import com.hazelcast.spi.properties.GroupProperty;

import java.util.HashMap;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.TimeUnit;

/**
 * Created by alarmnummer on 5/19/16.
 */
public class Main {

    private final static int instanceCount = 4;
    private final static int partitionCount = 12000;
    private final static int delayRangeMs = (int)TimeUnit.SECONDS.toMillis(30);

    public static void main(String[] args) {
        Config config = new Config();
        config.setProperty(GroupProperty.PARTITION_COUNT.getName(), "" + partitionCount);

        for (int k = 0; k < instanceCount; k++) {
            TestThread thread = new TestThread(config);
            thread.start();
        }
    }

    public static class TestThread extends Thread {
        private final Config config;

        public TestThread(Config config) {
            this.config = config;
        }

        public void run() {
            Random random = new Random();

            try {
                Thread.sleep(random.nextInt(delayRangeMs));

                HazelcastInstance hz = Hazelcast.newHazelcastInstance(config);
                for (; ; ) {
                    Thread.sleep(1000);

                    Map<Address, Integer> partitionsPerMember = new HashMap<Address, Integer>();
                    for (Partition partition : hz.getPartitionService().getPartitions()) {
                        Member owner = partition.getOwner();
                        if (owner == null) {
                            continue;
                        }

                        Integer value = partitionsPerMember.get(owner.getAddress());
                        Integer result = value == null ? 1 : value + 1;
                        partitionsPerMember.put(owner.getAddress(), result);
                    }

                    System.out.println(getName() + " " + partitionsPerMember);
                }
            } catch (Throwable t) {
                t.printStackTrace();
            }
        }
    }
}
```

The balance is being restored, but it is a slow process.

**Comment by mdogan** [comment]
I don't understand what is the issue here. You're starting instances in parallel and initializing partition distribution immediately. What's happening is, initials 1 or 2 members will own all the partitions and then migration/rebalance process will kick in after each new member joins and migration partition replicas.

**Comment by jerrinot** [comment]
@mdogan: the issue is that with 4 members and 12,000 partitions it takes quite a few minutes to rebalance. even there are no data stored.

**Comment by mdogan** [comment]
4 x 12000 = 48000 replicas. So, there will be around 36000 migration transactions to execute one by one. Each transaction involves invocations between 2~3 members and commit phase invocations to migration target + all cluster members. So, in total there will be around 7 invocations for each transaction.  That means 36000 x 7 = 252000 invocations.

**Comment by Donnerbart** [comment]
> You're starting instances in parallel and initializing partition distribution immediately.
Isn't there a property like "wait x seconds for other members before distributing the partitions for the first time" for setups like this?

**Comment by mdogan** [comment]
Partitions won't be distributed unless you touch them. So, if you touch a partition after all members are started then there will be no migration at all.

**Comment by Donnerbart** [comment]
So what's that triggering here? The `hz.getPartitionService().getPartitions()` call?

**Comment by mdogan** [comment]
`Member owner = partition.getOwner()`

**Comment by pveentjer** [comment]
The problem is that for +/- 20 minutes the system is in a totally unbalanced state and 'all' traffic is send to a single member. 

Even though there is not a lot of state to migrate; the migration process itself is time based. This is something which should be addressed. So as long as there is partitions to migrate, you migrate as fast as you can and as soon as there is no partitions to migrate, checking for migration is reduced to e.g. 1/second.

**Comment by pveentjer** [comment]
Another thing I have observed is that from time to time the migration seems to stall for some time. 

```
Mon May 23 08:49:07 EEST 2016 {[192.168.1.104]:5701=8951, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1019, [192.168.1.104]:5704=934}
Mon May 23 08:49:08 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:09 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:10 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:11 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
May 23, 2016 8:49:11 AM com.hazelcast.internal.partition.InternalPartitionService
INFO: [192.168.1.104]:5701 [dev] [3.7-SNAPSHOT] Remaining migration tasks in queue => 26607
Mon May 23 08:49:12 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:13 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:14 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:15 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:16 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:17 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:18 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:19 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:20 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:21 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:22 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:23 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:24 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:25 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:26 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
May 23, 2016 8:49:26 AM com.hazelcast.internal.partition.InternalPartitionService
INFO: [192.168.1.104]:5701 [dev] [3.7-SNAPSHOT] Remaining migration tasks in queue => 26010
Mon May 23 08:49:27 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:28 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:29 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:30 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:31 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:32 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:33 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:34 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:35 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:36 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:37 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:38 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:39 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:40 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:41 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
May 23, 2016 8:49:41 AM com.hazelcast.internal.partition.InternalPartitionService
INFO: [192.168.1.104]:5701 [dev] [3.7-SNAPSHOT] Remaining migration tasks in queue => 25413
Mon May 23 08:49:42 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:43 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:44 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:45 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:46 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:47 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:48 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:49 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:50 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:51 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:52 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:53 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:54 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:55 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:56 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
May 23, 2016 8:49:56 AM com.hazelcast.internal.partition.InternalPartitionService
INFO: [192.168.1.104]:5701 [dev] [3.7-SNAPSHOT] Remaining migration tasks in queue => 24816
May 23, 2016 8:49:56 AM com.hazelcast.internal.partition.InternalPartitionService
WARNING: [192.168.1.104]:5704 [dev] [3.7-SNAPSHOT] Master version should be greater than ours! Local version: 26243, Master version: 26242 Master: [192.168.1.104]:5701
Mon May 23 08:49:57 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:58 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:49:59 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:00 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:01 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:02 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:03 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:04 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:05 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:06 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:07 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:08 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:09 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:10 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:11 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
May 23, 2016 8:50:11 AM com.hazelcast.internal.partition.InternalPartitionService
INFO: [192.168.1.104]:5701 [dev] [3.7-SNAPSHOT] Remaining migration tasks in queue => 24224
May 23, 2016 8:50:11 AM com.hazelcast.internal.partition.InternalPartitionService
WARNING: [192.168.1.104]:5703 [dev] [3.7-SNAPSHOT] Master version should be greater than ours! Local version: 26834, Master version: 26833 Master: [192.168.1.104]:5701
Mon May 23 08:50:12 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:13 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:14 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:15 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:16 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:17 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:18 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:19 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:20 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:21 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:22 EEST 2016 {[192.168.1.104]:5701=8945, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=934}
Mon May 23 08:50:23 EEST 2016 {[192.168.1.104]:5701=8942, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=937}
Mon May 23 08:50:24 EEST 2016 {[192.168.1.104]:5701=8930, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=949}
Mon May 23 08:50:25 EEST 2016 {[192.168.1.104]:5701=8919, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=960}
Mon May 23 08:50:26 EEST 2016 {[192.168.1.104]:5701=8910, [192.168.1.104]:5703=1096, [192.168.1.104]:5702=1025, [192.168.1.104]:5704=969}
```

**Comment by mdogan** [comment]
Migration doesn't stall. See number of migration tasks:

```
INFO: [192.168.1.104]:5701 [dev] [3.7-SNAPSHOT] Remaining migration tasks in queue => 26607
INFO: [192.168.1.104]:5701 [dev] [3.7-SNAPSHOT] Remaining migration tasks in queue => 26010
INFO: [192.168.1.104]:5701 [dev] [3.7-SNAPSHOT] Remaining migration tasks in queue => 25413
INFO: [192.168.1.104]:5701 [dev] [3.7-SNAPSHOT] Remaining migration tasks in queue => 24816
INFO: [192.168.1.104]:5701 [dev] [3.7-SNAPSHOT] Remaining migration tasks in queue => 24224
```

You're only printing number of owned 0th replicas. But migration process handles migration/replication of all replicas, 0 to 6.

**Comment by pveentjer** [comment]
Ok. 

Is there any way we can speed up the migration process? And is there a simple way to verify that a system, when all nodes are up and running, is in stable state (so no migrations needed).

We need something like this for the simulator so that we only starts certain tests after the system is 'stable'. This is needed so that e..g key generation generates key with either local or remote behavior.

**Comment by mdogan** [comment]
Simply don't execute any partition aware task until all members are joined. 
I think (or was thinking) simulator's start phase initiates all members before switching to the warmup and run phases. If that's the case then you should not see any migration while running simulator tests.

**Comment by pveentjer** [comment]
The problem is not only for the simulator; I believe that problem can be fixed by setting initial cluster size. Till so far I have not seen any 'violations' due to partitions being migrated.

The problem is that for an end user application it is very easy to force partitions being used .e.g map.size. Once that happens, the partitions are materialized and it will take a lot of time when other member join the cluster that balance is restored.

**Comment by mdogan** [comment]
Solution is the same for end user too, not to trigger partition assignments until all expected members join or set initial cluster size. If you have 12k partitions and want safety, you should pay the cost. Problem is you want to access partitions before cluster is formed completely and also don't want any migrations to new coming members.

**Comment by pveentjer** [comment]
I understand your point.

For me it was really a WTF moment because it wasn't immediately obvious what the problem was. So if I can run into such an issue, an end user will for sure.

Is there no way we can speed up the migration? Especially if a partition is empty.

**Comment by mdogan** [comment]
Each migration is handled one by one by master node in a transactional manner. Before it was only for 0th replicas, now it's for all replicas. Former mode was closer to A, current one is C. 
Only owner of 0th replica knows if a partition is empty, so still master should ask owner for each migration. It cannot assign all replica owners beforehand (once partition table is initialized).

**Comment by mtopolnik** [comment]
Can migrations to the same node be batched in a single transaction?

**Comment by metanet** [comment]
Batching is not a very easy solution. We already complain about big network packages and batching partitions will make that problem worse. Checking size of a partition will also bring more complexity and coordination since the migration process is coordinated by master node and it will require more communication among nodes before starting a migration. 

Anyway, this issue has nothing to do with 3.7. We can discuss sensible improvements for 3.8 once it starts.

**Comment by mdogan** [comment]
Related to batching;
Each migration touches at least 2 endpoints; source and destination. Only the migrations from the same source to the same destination can be batched, otherwise handling failures will be very difficult. 
I'm not sure how much that will speedup the process.

Also as Basri said, generally a partition will contain fairly large data.

**Comment by gregrluck** [comment]
The existence of this open issue is causing confusion. I met someone who was confused by it. 

The normal use of Hazelcast is to start it then perform work on it.  As @mdogan  says: "Partitions won't be distributed unless you touch them. So, if you touch a partition after all members are started then there will be no migration at all."

In a production scenario with a huge number of partitions that is going to correspond to a 5-10 node cluster. When a node dies we would only have 1/5 or 1/10 of the partitions migrating, not 50% as in this contrived case.

So I think we should replace the contrived test showing a pathological case with a more realistic one, or close this issue.

Also, question: Would setting hazelcast.initial.min.cluster.size=4 prevent partition intialisation?

**Comment by wildnez** [comment]
In case of a failure, only backup partitions move within the cluster across nodes, not the primary partitions. To the original point in this thread, I think the current design makes more sense otherwise if we keep adding nodes after storing data in partitions in initial few nodes (this is the more common usage as no one just initialises a map and not put any data in it), the cluster will always be busy with repartitioning until all nodes have joined. 

There's a property called max.parallel.replication with default set to 5. Please correct if I'm wrong but this property can be used to modify number of threads on each node that participate in replication. This has worked well and tested in production deployments with 2k+ partitions storing 100s of million of records.

For 12k partitions, this problem does seem serious but not much if the cluster is used properly. As @mdogan had earlier said, this more about trade-off and a user needs to decide what they want.

**Comment by pveentjer** [comment]
@gregrluck hazelcast.initial.min.cluster.size=4 will wait with starting the system until this number of members are found. So no calls are made that are touching conditions and then you don't suffer from the slow connections.

I disagree that the test is a pathalogical case. The test is made to show the behavior.. But actually I'm seeing the problem in simulator. So when I create a cluster and do not pass the min cluster size option, then it takes a very long time for performance to become acceptable because in this time partitions are being migrated.

**Comment by mdogan** [comment]
Partition rebalance should be significantly faster with https://github.com/hazelcast/hazelcast/pull/14351 change in 3.12. 

In 4.1, we are planning to implement parallel migrations and to improve rebalance speed further.

**Comment by Holmistr** [comment]
I believe that we can close this issue as we have invested significant effort into speeding up migrations. Let's close it and possibly open another issue for particular case.
